File: src/lae/install/reload.py
"""Reload desktop components after install / uninstall."""

from __future__ import annotations

from lae.integrations import hyprland
from lae.util import subprocess as sp


def reload_hypr() -> bool:
    if not hyprland.available():
        return False
    hyprland.reload_config()
    return True


def restart_waybar() -> bool:
    if sp.which("omarchy-restart-waybar"):
        sp.run(["omarchy-restart-waybar"], check=False)
        return True
    if not sp.which("waybar"):
        return False
    sp.run(["pkill", "-9", "-x", "waybar"], check=False)
    sp.run(["setsid", "waybar"], check=False)
    return True


def apply_after_hypr() -> list[str]:
    actions: list[str] = []
    if reload_hypr():
        actions.append("reloaded Hyprland config")
    elif sp.which("hyprctl"):
        actions.append("Hyprland not active — run `hyprctl reload` after login")
    return actions


def apply_after_waybar() -> list[str]:
    actions: list[str] = []
    if restart_waybar():
        actions.append("restarted Waybar")
    elif sp.which("waybar"):
        actions.append("run `omarchy-restart-waybar` to apply Waybar changes")
    return actions

File: crates/tsk-core/src/install/reload.rs
use std::process::{Command, Stdio};

use crate::error::Result;

pub fn restart_waybar() -> bool {
    if command_exists("omarchy-restart-waybar") {
        let _ = Command::new("omarchy-restart-waybar")
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status();
        return true;
    }
    if command_exists("waybar") {
        let _ = Command::new("pkill")
            .args(["-9", "-x", "waybar"])
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status();
        let _ = Command::new("setsid")
            .args(["waybar"])
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status();
        return true;
    }
    false
}

pub fn apply_after_hypr() -> Result<Vec<String>> {
    let mut actions = Vec::new();
    if command_exists("hyprctl") {
        let _ = Command::new("hyprctl")
            .arg("reload")
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status();
        actions.push("reloaded Hyprland config".into());
    } else {
        actions.push("Hyprland not active — run `hyprctl reload` after login".into());
    }
    Ok(actions)
}

pub fn apply_after_waybar() -> Vec<String> {
    if restart_waybar() {
        vec!["restarted Waybar".into()]
    } else if command_exists("waybar") {
        vec!["run `omarchy-restart-waybar` to apply Waybar changes".into()]
    } else {
        Vec::new()
    }
}

/// Apply pending Hyprland and Waybar integration changes once.
pub fn apply_after_install(reload_hypr: bool, reload_waybar: bool) -> Result<Vec<String>> {
    let mut actions = Vec::new();
    if reload_hypr {
        actions.extend(apply_after_hypr()?);
    }
    if reload_waybar {
        actions.extend(apply_after_waybar());
    }
    Ok(actions)
}

fn command_exists(name: &str) -> bool {
    Command::new("sh")
        .arg("-c")
        .arg(format!("command -v {name}"))
        .output()
        .is_ok_and(|o| o.status.success())
}
